src/redownload/web_parsing.py

```diff
@@ -0,0 +1,58 @@
+""" The module for HTML parsing related functions. """
+import urllib.request
+
+import bs4
+
+from . import exceptions
+
+
+def download_from_url(url: str) -> bs4.BeautifulSoup:
+    """Downloads an HTML page from a url and converts it to a BeautifulSoup object.
+
+    :param url: The URL to download HTML from, in a string.
+    :return: BeautifulSoup object extracted from the url.
+    """
+    request = urllib.request.urlopen(url)
+    html_doc = request.read()
+    html_soup = bs4.BeautifulSoup(html_doc, features="html.parser")
+    return html_soup
+
+
+def extract_links(page: bs4.BeautifulSoup, extensions: list = None) -> list:
+    """Extracts all the links with a file extension listed in the extensions param from a BeautifulSoup object and
+    returns a list of them. Raises an exception if there are no audio links in the object. If extensions is not
+    specified, returns all links on the page.
+
+    :param page: a BeautifulSoup object to extract audio links from.
+    :param extensions: OPTIONAL: a list of file extensions to filter for. If not specified, returns all links.
+    :return: a list of links to audio files ending in .flac or .mp3
+    """
+    all_links = []
+    correct_links = []
+    # Get all links on page.
+    for link in page.findAll("link"):
+        all_links.append(link.get("href"))
+    for link in page.findAll("a"):
+        all_links.append(link.get("href"))
+
+    # FIXME - Somehow there are None objects at the end of the list.
```

Comment:
Maybe fix this before merging?

Reply:
The code directly under it is a band-aid fix because I have no idea what's actually causing them.

Reply:
The problem probably results from the usage of link.get() Usually get() will return None if the requested key is not in the data structure. I don't know why a link wouldn't have a href but that is what's causing your problem. Maybe you already knew that IDK

Reply:
Looking now, I think it's because it's also getting `<a>` tags, which can have no href and just act as an anchor. Would using `link["href"]` and handling the exception solve this?

Reply:
Yeah you can either handle the exception or check if `"href" in link`

Reply:
You can't check `if "href" in link` because it's not a dictionary or a list, but I added a check to make sure objects being added are not null before they get added to the list. It also prevents duplicates in the same if statement.